### PR TITLE
DM-32442: Add Java unit tests for SAL.

### PR DESCRIPTION
**Pull request by wvreeven**

### CODE REVIEW

**Review by dmills-lsst — COMMENTED** [review]
So far so good...
BaseTestCase.java references SALPY in comment
LsstDdsQosTest,java has /home/wreeven path
Does @throws Exception In case of a problem related to reflection mean the tests do not work?

File: java_tests/src/test/java/org/lsst/sal/SAL_ScriptTest.java
Comment (r-owen) [inline java_tests/src/test/java/org/lsst/sal/SAL_ScriptTest.java:8]: The way SAL handles generic topics has changed, making this comment and the test names and details somewhat out of date. The XML field name has changed from Generics to AddedGenerics and "no" is no longer a valid entry. But this is fundamentally a test of the way that SAL handles generic topics, and Script is still unusual in not being a CSC.

Please at least update the code comment, test class name and test name(s) to reflect the current reality.

Please also consider updating the test itself, e.g.:
* Test that `Script` has no `summaryState` event nor `enable` command (two topics that all CSCs have). This is a useful change from testing for the `enterControl` command, because now very CSCs have the `enterControl` command so it is normal for it to be missing.
* Test that `Test` has no `enterControl` command.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (r-owen) [inline java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:34]: Surely Dave is willing to add getLeapSeconds to Java?

I'd be a lot more comfortable if these issues you are finding in the unit tests get fixed in SAL. Dave usually fixes things quickly and if anything he can fix quickly would be a big win over committing disabled tests.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (r-owen) [inline java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:52]: "ran" -> "run"

This is getting really scary. This is a very important test for serious problems we have had. Please try to find some way to kill the JVM between test invocations.

Perhaps using a different LSST_DDS_DOMAIN_PREFIX for each test invocation will fix this particular issue?

File: java_tests/src/test/java/org/lsst/sal/BaseTestCase.java
Comment (r-owen) [inline java_tests/src/test/java/org/lsst/sal/BaseTestCase.java:26]: Please generate a random LSST_DDS_DOMAIN_PREFIX env var before constructing the remote and controller. That is potentially useful for keeping tests from colliding with each other. I think it won't be a problem that the JVM caches information, because SAL reads the env var when it creates the `SAL_<name>` object.

File: java_tests/src/test/java/org/lsst/sal/SAL_ScriptTest.java
Comment (r-owen) [reply java_tests/src/test/java/org/lsst/sal/SAL_ScriptTest.java:8]: If you prefer, I can update the Python code first.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (r-owen) [inline java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:49]: What does `reflection` mean in this context? Please expand this comment (and the many others like it).

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (r-owen) [inline java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:294]: The ability to read late-joiner events is absolutely critical. It feels like a very large and important swath of SAL is not tested due to this JVM issue.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (r-owen) [inline java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:289]: "can an event" -> "can read the most recent event" or similar.

**Review by r-owen — CHANGES_REQUESTED** [review]
This is a great start, but with some worrying limitations:
* Some important tests are disabled due to a JVM issue. I think it is important to try to figure out how to run these (though I won't block this ticket on that basis as long as you file a ticket to fix this problem).
* Other tests are disabled because SAL itself is apparently misbehaving. For these I think it is important to ask Dave to fix the problems, and if he can do it quickly, then I would strongly prefer to hold off merging this until the problems are fixed and the tests can be enabled.

It is possible that a few of the JVM-related issues can be worked around by setting a random LSST_DDS_DOMAIN_PREFIX before constructing each pair of SAL objects. It is well worth trying, especially as it prevents interference with production code. (If you are doing that, I missed it)

File: doc/version_history.rst
Comment (wvreeven) [reply doc/version_history.rst:47]: Good point! Not sure why I didn't realize that myself. Fixed.

File: java_tests/src/test/java/org/lsst/sal/BaseTestCase.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/BaseTestCase.java:27]: I made sure that it always is positive.

File: java_tests/src/test/java/org/lsst/sal/BaseTestCase.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/BaseTestCase.java:59]: Done.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:54]: I think it is yes. I'll discuss this with Dave and will see if we can fix this or open a new ticket to fix this in sal and then fix the unit test as well.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:132]: I added them to make sure that it was clear that the python tests were implemented as literally as possible. That goal was achieved and I have removed these lines.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:161]: I'll discuss this with Dave and will see if we can fix this or open a new ticket to fix this in sal and then fix the unit test as well.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:186]: Done

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:189]: I'll discuss this with Dave and will see if we can fix this or open a new ticket to fix this in sal and then fix the unit test as well.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:234]: I have literally implemented the Python tests and have not checked if the sleeps are necessary or not. OK if I move checking that and removing the unnecessary sleeps in another ticket?

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:266]: I'll discuss this with Dave and will see if we can fix this or open a new ticket to fix this in sal and then fix the unit test as well.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:292]: Done

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:318]: Java has MANY checked exceptions that require the developer to either surround a statement with a try/catch block or declare that the calling method throws the exception. The methods that get called in the methods declare to throw checked exceptions, which is why they need to be either caught or re-thrown. The pattern applied here is the accepted pattern for JUnit to make sure that an exception is thrown by the method being called while still being able to continue in the method. I've added a comment briefly explaining that but any Java developer is supposed to know how to handle checked exceptions.

File: java_tests/src/test/java/org/lsst/sal/SAL_ScriptTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/SAL_ScriptTest.java:8]: Please update the Python code first. Unfortunately I am very busy with the dome software installation at the summit and therefore prefer to not have to spend too much time on this ticket if I can possibly help it.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:34]: The way Java handles leap seconds probably is such that this cannot be fixed by Dave. Java basically makes a smooth transition by enlarging the time span of a second during a certain period of time so no "additional" second is needed. It is very ugly and bad, but that's the way it is. If you prefer I can remove the test case.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:49]: Reflection is a Java standard present in the JDK and JVM for introspection of classes and class instances. Much like getattr in Python. Any Java developer knows what reflection means in a Java context so from a Java point of view there is no need to explain this.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:289]: Again, I implemented the python tests as literal as possible meaning that I copied the python doc strings as well. I've fixed this in both the python and the Java files.

File: java_tests/src/test/java/org/lsst/sal/LsstDdsQosTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/LsstDdsQosTest.java:16]: Thanks for this comment. It made me think and it actually turns out that that is possible. The JVM in which the unit tests are run can be forked such that that actual unit test code itself runs in a separate JVM. And it also is possible to make sure that a forked JVM is not reused, meaning that each unit test class starts with a fresh JVM. I opened DM-32914 for this.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:52]: See my previous remark about forking the JVM when running the unit tests. This issue will be addressed in DM-32914.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:294]: See my previous remark about forking the JVM when running the unit tests. This issue will be addressed in DM-32914.

File: java_tests/src/test/java/org/lsst/sal/BaseTestCase.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/BaseTestCase.java:26]: Would it be OK if I addressed this as well in DM-32914?

**Comment by wvreeven** [comment]
Thanks for all the useful comments. Regarding the question raised by Dave "Does @throws Exception In case of a problem related to reflection mean the tests do not work?", the answer is no. Java has the concept of checked exceptions which mean that you either need to try/catch the exception or declare that your method throws the exception for the compiler to be happy. I chose the latter since we know that the class members, for which the values are being set, exist and are accessible (meaning they are public and not private).

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java]: Fixed.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:149]: I have opened DM-32952 for this.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:173]: I have opened DM-32952 for this.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:218]: I have opened DM-32952 for this.

File: java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java
Comment (wvreeven) [reply java_tests/src/test/java/org/lsst/sal/ErrorHandlingTest.java:250]: I have opened DM-32952 for this.

File: java_tests/src/test/java/org/lsst/sal/SAL_ScriptTest.java
Comment (r-owen) [reply java_tests/src/test/java/org/lsst/sal/SAL_ScriptTest.java:8]: I filed DM-32956 to fix the Python and implemented it in https://github.com/lsst-ts/ts_sal/pull/168.

I would rather have this bit of Java cleaned up before merging, even if it adds a delay while you deal with more important things. I am willing to do the cleanup myself, if need be.

File: java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java
Comment (r-owen) [reply java_tests/src/test/java/org/lsst/sal/SAL_TestTest.java:34]: This seems very surprising. The value returned by SAL's time function (which is not this one) is the value of CLOCK_TAI. I don't think Java would change that. I don't know how ts_sal computes the value returned by getLeapSeconds, but again it should be easy to get that value through to Java. In any case I've started a discussion on Slack to learn more about this and we can revisit this based on that discussion.

**Review by r-owen — CHANGES_REQUESTED** [review]
Thanks for filing the new tickets. A few things remain:
* I think the version number needs a smaller jump.
* I would still like to see the AddedGenerics test updated as per my suggestion. I'm willing to do this if you don't have time. But I think it would be very quick. I updated the python code on https://github.com/lsst-ts/ts_sal/pull/168 if you want an example (though I cannot yet it yet because I can't build the Script SALPY library; I've filed a ticket about that).
* I am not yet convinced by the argument about getLeapSeconds, though that can be postponed to one of the other tickets.

**Review by r-owen — APPROVED** [review]
With @wvreeven 's kind permission I updated the generics test for modern AddedGenerics tag (matching a very recent update to the SALPY unit tests) and rebased the PR.

I also changed the version number to 6.1.0.

Please run the tests to make sure I didn't break anything. Feel free to merge once you are happy, and push back if you don't like my changes.

One more suggestion to @wvreeven and @dmills-lsst: please consider not releasing ts_sal 6.1 until the related tickets are done, so that the Java issues you discovered (that required disabling tests) are fixed, and the Java tests are running (to keep these issues fixed). There's no rush and new SAL versions tend to be a big deal.

File: doc/version_history.rst
Comment (wvreeven) [reply doc/version_history.rst:12]: I lowered it to 6.1.0

**Comment by wvreeven** [comment]
I adopted the doc strings for the checkXGetPutRaises in the Java tests code.
I adopted the python changes to AddedGenerics in the SAL_ScriptTest.java code.
I added a crude getLeapSecondsTest which verifies that the returned value for leap seconds is greater than or equal to zero.

Also, I modified the next version for sal from 6.0.1 (as added by Russell) to 6.1.0.

**Pull request merged**
